### PR TITLE
Service Discovery: allow multiple schemes to be specified in request

### PR DESCRIPTION
**Pull request by ReubenBond**
This PR adds support in Service Discovery for specifying multiple schemes during resolution, separated by a `+` character. This allows developers to enforce HTTPS in production while falling back to HTTP during development and testing.
For example, a user can allow both HTTPS and HTTP like so:

```csharp
builder.Services.AddHttpClient<CatalogServiceClient>(
  c => c.BaseAddress = new("https+http://catalogservice"));
```

The order of schemes is important: if endpoints are found matching the first scheme, no endpoints from subsequent schemes are allowed.

This PR also introduces configuration to specify an allow-list of acceptable schemes for service discovery:

For example, to allow only `https://`:
```csharp
builder.Services.Configure<ServiceDiscoveryOptions>(options =>
{
    options.AllowedSchemes = ["https"];
});
```
To allow `http://` or `https://`:
```csharp
builder.Services.Configure<ServiceDiscoveryOptions>(options =>
{
    options.AllowedSchemes = ["https", "http"];
});
```
To allow any scheme (the default):
```csharp
builder.Services.Configure<ServiceDiscoveryOptions>(options =>
{
    options.AllowedSchemes = ServiceDiscoveryOptions.AllSchemes;
});
```

This PR changes the format of Service Discovery configuration so that the endpoint name is included in the configuration path, if an endpoint name is set. If no endpoint name is specified, then the scheme is used (and again, schemes are checked in the specified order, with only the first present being selected). If no endpoint name or scheme is specified, `"default"` is used as the endpoint name.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2719)

### CODE REVIEW

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs [hunk 1]
```diff
@@ -243,19 +243,21 @@ public void WriteServiceDiscoveryEnvironmentVariables(IResource resource)
                     ? endpointReferenceAnnotation.Resource.Annotations.OfType<EndpointAnnotation>().Select(sb => sb.Name)
                     : endpointReferenceAnnotation.EndpointNames;
 
-                var endpointAnnotationsGroupedByScheme = endpointReferenceAnnotation.Resource.Annotations
```
Comment (davidfowl) [inline src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:246]: Can we delete all of this code and rely on the `CreateEndpointReferenceEnvironmentPopulationCallback`? Now that we're setting EndpointReference directly, that should handle the formatting.

File: src/Aspire.Hosting/Extensions/ResourceExtensions.cs [hunk 2]
```diff
@@ -94,9 +94,18 @@ public static bool TryGetEndpoints(this IResource resource, [NotNullWhen(true)]
     /// <param name="allocatedEndPoints">When this method returns, contains the allocated endpoints for the specified resource, if they exist; otherwise, <c>null</c>.</param>
     /// <returns><c>true</c> if the allocated endpoints were successfully retrieved; otherwise, <c>false</c>.</returns>
     [Obsolete("Use GetEndpoints instead.")]
-    public static bool TryGetAllocatedEndPoints(this IResource resource, [NotNullWhen(true)] out IEnumerable<AllocatedEndpointAnnotation>? allocatedEndPoints)
+    public static bool TryGetAllocatedEndPoints(this IResource resource, [NotNullWhen(true)] out IEnumerable<AllocatedEndpoint>? allocatedEndPoints)
```
Comment (davidfowl) [inline src/Aspire.Hosting/Extensions/ResourceExtensions.cs:97]: Delete this method

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 3]
```diff
@@ -309,15 +309,10 @@ private static void AddAllocatedEndpointInfo(IEnumerable<AppResource> resources)
                     throw new InvalidOperationException($"Service '{svc.Metadata.Name}' needs to specify a port for endpoint '{sp.EndpointAnnotation.Name}' since it isn't using a proxy.");
                 }
 
-                var a = new AllocatedEndpointAnnotation(
-                    sp.EndpointAnnotation.Name,
-                    PortProtocol.ToProtocolType(svc.Spec.Protocol),
+                sp.EndpointAnnotation.AllocatedEndpoint = new AllocatedEndpoint(
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:312]: clean

File: tests/Aspire.Hosting.Tests/WithReferenceTests.cs [hunk 4]
```diff
@@ -93,13 +93,16 @@ public async Task ResourceWithConflictingEndpointsProducesAllEnvironmentVariable
...
 
-        // Create a binding and its metching annotation (simulating DCP behavior)
+        // Create a binding and its matching annotation (simulating DCP behavior)
         testProgram.ServiceABuilder.WithHttpsEndpoint(1000, 2000, "mybinding");
         testProgram.ServiceABuilder.WithEndpoint("mybinding", e => e.AllocatedEndpoint = new AllocatedEndpoint(e, "localhost", 2000));
 
         testProgram.ServiceABuilder.WithHttpsEndpoint(1000, 3000, "mybinding2");
         testProgram.ServiceABuilder.WithEndpoint("mybinding2", e => e.AllocatedEndpoint = new AllocatedEndpoint(e, "localhost", 3000));
 
+        // The appsettings.json adds an "http" endpoint
```
Comment (DamianEdwards) [inline tests/Aspire.Hosting.Tests/WithReferenceTests.cs:103]: I think you mean `launchSettings.json`?

Comment (ReubenBond) [reply tests/Aspire.Hosting.Tests/WithReferenceTests.cs:103]: yes, thanks

**Review by davidfowl — APPROVED** [review]

**Pull request merged**